grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -735,3 +686,88 @@ def _GetTypeName(
     return cast(str, primitive_types[cls]["name"])
 
   return str(cls)  # Cover `BinaryStream` and `None`.
+
+
+def _GetReferenceObject(type_name: str) -> SchemaReference:
+  """Get a `Reference Object` that points to a schema definition.
+
+  All types (including protobuf primitives) are expected to have been
+  previously defined in the `components` field of the root `OpenAPI Object`
+  and are used via OpenAPI references.
+
+  Args:
+    type_name: The name of the type for which an OpenAPI `Reference Object`
+      will be created and returned.
+
+  Returns:
+    An OpenAPI `Reference Object` representing the path to the actual OpenAPI
+    schema definition of the selected type.
+  """
+  return {
+    "$ref": f"#/components/schemas/{type_name}"
+  }
+
+
+def _GetArraySchema(items_type_name: str) -> ArraySchema:
+  """Get the schema of an array with items of the given type."""
+
+  return {
+    "type": "array",
+    "items": _GetReferenceObject(items_type_name)
+  }
+
+
+def _GetFieldSchema(
+    field_descriptor: FieldDescriptor
+) -> Union[SchemaReference, ArraySchema, OneofFieldSchema]:
+  """Create a dictionary to be used as the schema of a message field.
+
+  This function takes into consideration the fact that the field might be part
+  of a `protobuf.oneof` and wraps the `Reference Object` or the OpenAPI array
+  accordingly in order to include a description specifying the name of the
+  `protobuf.oneof`.
+
+  Args:
+    field_descriptor: The protobuf `FieldDescriptor` associated with the target
+      field.
+
+  Returns:
+    If the field is not part of a `protobuf.oneof`, then a simple `Reference
+    Object`, else, a dictionary that includes a description entry along the
+    `Reference Object`.
+  """
+  # The semantic of `protobuf.oneof` is not currently supported by the
+  # OpenAPI Specification. See this GitHub issue [1] for more details.
+  #
+  # [1]: github.com/google/grr/issues/822
+  containing_oneof: OneofDescriptor = field_descriptor.containing_oneof
+  if containing_oneof is None:
+    if field_descriptor.label == protobuf2.LABEL_REPEATED:
+      return _GetArraySchema(_GetTypeName(field_descriptor))
+
+    return _GetReferenceObject(_GetTypeName(field_descriptor))
+  else:
```

Comment:
I think you can remove this `else` to reduce nesting.

Reply:
Thank you! Removed it.